src/binary-writer.cc

```diff
@@ -1188,8 +1183,17 @@ Result BinaryWriter::WriteModule() {
       Offset body_size_offset =
           WriteU32Leb128Space(leb_size_guess, "func body size (guess)");
       WriteFunc(func);
-      WriteFixupU32Leb128Size(body_size_offset, leb_size_guess,
+      auto func_start_offset = body_size_offset - last_section_payload_offset_;
+      auto func_end_offset = stream_->offset() - last_section_payload_offset_;
+      auto delta = WriteFixupU32Leb128Size(body_size_offset, leb_size_guess,
                               "FIXUP func body size");
+      if (current_reloc_section_ && delta != 0) {
+        for (Reloc& reloc : current_reloc_section_->relocations) {
+          if (reloc.offset >= func_start_offset && reloc.offset <= func_end_offset) {
+            reloc.offset += delta;
+          }
+        }
+      }
```

Comment:
The complexity needed to handle these adjustments is a little sad.  I wonder if there is a better way we can do this.  I can think of a few options:

1. Write a patchable 5-byte LEB here
2. Have some way of precalualting the size of a function.
3. WriteFunc to a temporary location and them memmove into place once we know the size?

I'm not suggesting these for this PR but perhaps for future works.

I can't remember why we didn't choose any of these originally.  @binji what do you think?

Reply:
Considering that we use patchable 5-byte LEBs in relocatable code anyways and it gets optimized later, I think that might be the easiest and most sensible solution.

Reply:
Yeah, the complexity here is annoying. I want to make sure we don't write a larger size (just in case optimization doesn't happen), but perhaps we can do it by default when the output is relocatable.

Reply:
Yes exactly we would only write wide-LEBs for relocatable output which will be used as input wasm-ld

Reply:
If you want I can do that in this PR, it seems kinda pointless to introduce complexity just to remove it right afterwards.

Reply:
I'm fine with landing this PR as is.  It doesn't really add much.. just move the existing loop and adds a condition.

We can reason about using wide LEBs as a followup.

Reply:
although it might also make sense to just make a new PR for that and close this one, because it is a bit broader in scope.